core/src/main/java/io/questdb/cairo/O3PurgeDiscoveryJob.java

```diff
@@ -28,181 +28,266 @@
-                            partitionBy,
-                            partitionTimestamp,
-                            txnScoreboard,
-                            nameTxnToRemove,
-                            minTxnToExpect
-                    )) != 0) {
-                        // queue the job
-                        if (purgePubSeq != null) {
-                            long cursor = purgePubSeq.next();
-                            if (cursor > -1) {
-                                LOG.error()
-                                        .$("queuing [table=").$(tableName)
-                                        .$(", ts=").$ts(partitionTimestamp)
-                                        .$(", txn=").$(nameTxnToRemove)
-                                        .$(", errno=").$(errno)
-                                        .$(']').$();
-                                O3PurgeTask task = purgeQueue.get(cursor);
-                                task.of(
-                                        tableName,
-                                        partitionBy,
-                                        txnScoreboard,
-                                        partitionTimestamp,
-                                        nameTxnToRemove,
-                                        minTxnToExpect
-                                );
-                                purgePubSeq.done(cursor);
-                            } else {
-                                LOG.error()
-                                        .$("purge queue is full [table=").$(tableName)
-                                        .$(", ts=").$ts(partitionTimestamp)
-                                        .$(", txn=").$(nameTxnToRemove)
-                                        .$(", errno=").$(errno)
-                                        .$(']').$();
-                            }
-                        } else {
-                            LOG.error()
-                                    .$("could not purge [table=").$(tableName)
-                                    .$(", ts=").$ts(partitionTimestamp)
-                                    .$(", txn=").$(nameTxnToRemove)
-                                    .$(", errno=").$(errno)
-                                    .$(']').$();
-                        }
+        // find duplicate partitions
+        assert partitionList.size() % 2 == 0;
+        Vect.sort128BitAscInPlace(partitionList.getAddress(), partitionList.size() / 2);
+
+        long partitionTs = Numbers.LONG_NaN;
+        int lo = 0;
+        int n = (int) partitionList.size();
+
+        path.of(root).concat(tableName);
+        int tableRootLen = path.length();
+        try {
+            txnScoreboard.ofRO(path);
+            path.trimTo(tableRootLen);
+            txReader.ofRO(path, partitionBy);
+            loadLastTx(txReader);
+
+            for (int i = 0; i < n; i += 2) {
+                if (partitionList.get(i + 1) != partitionTs) {
```

Comment:
Double list access in a loop

Reply:
fixed in fixed in https://github.com/questdb/questdb/pull/1778/commits/c07fd7aa136747a7679d3865ffbb164535d0ee3d